### PR TITLE
SWC-6390: Changes to support 2FA Login

### PR DESCRIPTION
**Pull request by nickgros**
- add types/SynapseClient functions for 2fa login
- refactor `SynapseClient.detectSSOCode` to become `useDetectSSOCode` and add tests

### CODE REVIEW

File: packages/synapse-react-client/test/lib/utils/SynapseClient.test.ts [hunk 1]
```diff
@@ -71,17 +75,17 @@ describe('SynapseClient tests', () => {
           setCurrentAsyncStatus,
         )
 
-        expect(requestCaptor).toBeCalledTimes(1)
-        expect(requestCaptor).toBeCalledWith(
+        expect(requestCaptor).toHaveBeenCalledTimes(1)
```
Comment (nickgros) [inline packages/synapse-react-client/test/lib/utils/SynapseClient.test.ts:78]: ESLint is configured to prefer this name of this assertion (same behavior)

File: packages/synapse-react-client/demo/containers/Demo.tsx [hunk 2]
```diff
@@ -90,7 +91,8 @@ class Demo extends React.Component<DemoProps, DemoState> {
   public componentDidMount() {
     // Note:  All portals should do this once on the initial app load.
     // This looks for the access token cookie (HttpOnly, unable to directly access), and initialize the session if it does exists.
-    SynapseClient.detectSSOCode()
+    // TODO: Convert to function component to use this hook. Until then, SSO will not work.
+    // useDetectSSOCode()
```
Comment (nickgros) [inline packages/synapse-react-client/demo/containers/Demo.tsx:95]: Deprecated demo

File: packages/synapse-react-client/package.json [hunk 3]
```diff
@@ -162,7 +162,7 @@
     "@vitejs/plugin-legacy": "^3.0.2",
     "@vitejs/plugin-react": "^3.0.1",
     "@welldone-software/why-did-you-render": "^7.0.1",
-    "babel-jest": "^28.1.1",
+    "babel-jest": "^29.4.3",
```
Comment (nickgros) [inline packages/synapse-react-client/package.json:165]: Was running into what I think was an incompatibility between Jest 29.x and babel-jest 28.x that manifested as a stack overflow error. Issue went away when I aligned them

File: packages/synapse-react-client/src/lib/utils/hooks/useDetectSSOCode.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+} from '../SynapseClient'
+import { useEffect } from 'react'
+
+/*
+ * During SSO login, the authorization provider (Google) will send the user back to the portal with an authorization code,
+ * which can be exchanged for a Synapse user session. This function should be called whenever the root App is initialized
+ * (to look for this code parameter and complete the round-trip). If state is included, then we assume that this is being
+ * used for account creation, where we pass the username through the process.
+ */
+export default function useDetectSSOCode(
```
Comment (nickgros) [inline packages/synapse-react-client/src/lib/utils/hooks/useDetectSSOCode.ts:23]: No changes to this function yet to support 2FA.

File: packages/synapse-react-client/src/lib/utils/SynapseClient.ts [hunk 5]
```diff
@@ -1557,96 +1605,6 @@ export const getPrincipalAliasRequest = (
...
-/*
-During SSO login, the authorization provider (Google) will
-send the user back to the portal with an authorization code,
-which can be exchanged for a Synapse user session.
-This function should be called whenever the root App is initialized
-(to look for this code parameter and complete the round-trip).
-If state is included, then we assume that this is being used for account creation,
-where we pass the username through the process.
-*/
-export const detectSSOCode = (
```
Comment (nickgros) [inline packages/synapse-react-client/src/lib/utils/SynapseClient.ts:1569]: Moved and converted to custom hook

**Review by jay-hodgson — APPROVED** [review]

**Pull request merged**